### ISSUE TITLE
proto_library should use the released binary of protoc instead of compiling it from source.

### ISSUE THREAD

**Issue by ribrdb** [bazelbuild/rules_proto#35]
It's very annoying to have to wait to rebuild protoc when I haven' made any changes to it.

**Comment by Yannic** [comment]
Thanks!

The tricky part here is that this could result in generating with protoc version x and using the proto runtime from version y, which may or may not be supported by the language (e.g. Java-Lite doesn‘t [1]).

I don‘t think we will implement this short term, but let‘s keep this open as tracking issue.

[1] https://github.com/protocolbuffers/protobuf/tree/master/java#compatibility-notice

**Comment by Yannic** [comment]
Oh, wait. I was thinking of looking for a system-installed version of protoc.
Downloading a precompiled binary probably does not have the issue mentioned above.

**Comment by ribrdb** [comment]
I tried implementing this here: https://github.com/bazelbuild/rules_proto/compare/master...ribrdb:binaries

It seems to work sometimes. But java_proto_library still triggers building protoc and I can't figure out why.

**Comment by Yannic** [comment]
Awesome, I‘ll check it out next week. Would you mind creating a (draft) PR for it so we can have a discussion on the code?

I also need to think a bit about whether it would be better delaying this until after we‘ve finished migrating core proto rules to Starlark. I‘d rather not land things now just to revert them in a few month...

**Comment by dmivankov** [comment]
Tried the PR, didn't quite work for me
```
git_repository(
    name = "bazelbuildd_rules_proto",
    remote = "https://github.com/ribrdb/rules_proto.git",
    commit = "2c3dbcc39f6bb66068f4cc2432c674ac79cec47b"
)
load("@bazelbuildd_rules_proto//proto:repositories.bzl", "rules_proto_dependencies", "rules_proto_toolchains")
rules_proto_dependencies()
rules_proto_toolchains()
load("@bazelbuildd_rules_proto//proto:defs.bzl", "proto_library")
===
INFO: Call stack for the definition of repository 'com_google_protobuf' which is a protobuf_workspace (rule definition at /home/divankov/.cache/bazel/_bazel_divankov/010f6cf54aedc6c2912331bbadc8b1e4/external/bazelbuildd_rules_proto/proto/private/dependencies.bzl:158:22):
 - /home/divankov/.cache/bazel/_bazel_divankov/010f6cf54aedc6c2912331bbadc8b1e4/external/bazelbuildd_rules_proto/proto/repositories.bzl:26:5
- ... WORKSPACE location ...
ERROR: An error occurred during the fetch of repository 'com_google_protobuf':
   Not a regular file: /home/divankov/.cache/bazel/_bazel_divankov/010f6cf54aedc6c2912331bbadc8b1e4/external/rules_proto/proto/private/BUILD.release
```

**Comment by mleonhard** [comment]
Compiling protoc adds about 8 minutes to our CI build time.  This adds an extra eight minutes of downtime to any outage that requires a code fix.

**Comment by dmivankov** [comment]
Adding a cache to builds can help with that, but that's of course a bunch of extra setup

**Comment by dmivankov** [comment]
Ok, "not a regular file" was due to this PR assuming repository will be consumed under rules_proto name (@rules_proto//proto/private:BUILD.release reference)

Next up was
```
'external/com_google_protobuf_protoc_linux_x86_64/bin/protoc' in inputs. A tool is an input with executable=True set. All tools should be passed using the 'tools' argument instead of 'inputs' in order to make their runfiles available to the action. This safety check will not be performed once the action is modified to take a 'tools' argument. To temporarily disable this check, set --incompatible_no_support_tools_in_action_inputs=false.
```
The fix is to replace srcs with tools in _protoc_binary_impl

There's also a typo in proto/private/dependencies.bzl "com_google_protobuf_protoc_windows_x86-64" should be "com_google_protobuf_protoc_windows_x86_64"


But then tried to do grpc for java and it needs to recompile anyway :D
Next up is gRPC https://github.com/rules-proto-grpc/rules_proto_grpc/blob/master/java/BUILD.bazel going to https://github.com/grpc/grpc-java/blob/master/compiler/BUILD.bazel which wants "@com_google_protobuf//:protoc_lib" and wants to build it with https://github.com/grpc/grpc-java/tree/master/compiler/src/java_plugin/cpp (maybe there's a better/another way to do it, or of course this extended version could be binary-released and consumed in grpc java bazel files)

**Comment by kgreenek** [comment]
One large drawback to distributing protoc as a pre-built binary is that it is currently not possible to determine the host architecture from the WORKSPACE file.

This means that you couldn't easily determine which protoc binary to download (e.g. x86_64, aarch64, windows, etc.), and you'll likely break some of those architectures.

Here is an example of a project that went down this road and got bitten:
https://github.com/bazelbuild/rules_docker/issues/1155

In my case, we have CI set up to build for x86_64 and aarch64. We deploy to nvidia embedded platforms that use aarch64. This would also affect users who deploy to raspberry pi. I was able to employ the workaround from that issue with some pain, using a custom rule and hosting the aarch64 binary myself. Please be aware of this if you decide to distribute pre-built binaries that need to be downloaded in the WORKSPACE file.

**Comment by Yannic** [comment]
@kgreenek Thanks for bringing this up! @ribrdb PR (#36) works around this by declaring one workspace per architecture and uses `select()` to choose the binary during analysis/configuration (as opposed to `rules_docker`, which commits to a binary in the loading phase). If no pre-built `protoc` is available, `protoc` is built from source.

Does that answer your concerns?

**Comment by kgreenek** [comment]
Ah I see. That works for me. And I see aarch64 is supported too. Thanks!

**Comment by alexeagle** [comment]
Can this be closed since #36 was merged?

**Comment by Yannic** [comment]
> Can this be closed since #36 was merged?

Yes. Thank you for working on this @ribrdb!
